### ISSUE TITLE
CMS: when signing&encrypting data, can the flags be combined?

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#25491]
In PR #25132 the following new rather general function is going to be introduced:
```
CMS_EnvelopedData *CMS_sign_encrypt(BIO *data, X509 *sign_cert, STACK_OF(X509) *certs,
                                    EVP_PKEY *sign_key, unsigned int sign_flags, 
                                    STACK_OF(X509) *enc_recip, const EVP_CIPHER *cipher,
                                    unsigned int enc_flags, OSSL_LIB_CTX *libctx, const char *propq)
```
When reviewing this, [I placed a question](https://github.com/openssl/openssl/pull/25132#discussion_r1734722295):
> Maybe `sign_flags` and `enc_flags` can be merged into a single `flags` parameter, used for both?
Yet I don't have sufficient experience with the CMS API whether this makes sense.
@slontis or @mattcaswell what do you think?

which apparently went unnoticed, so now bringing it up more prominently here.

**Comment by DDvO** [comment]
It looks like to me that at least the `CMS_BINARY` flag would be given either both for signing and encrypting or for neither of them.

**Comment by DDvO** [comment]
BTW, is the proposed name `CMS_sign_encrypt()` still acceptable (for consistency with existing CMS functions)?
Or do we feel obliged to use `OSSL_CMS_sign_encrypt()`?

**Comment by mattcaswell** [comment]
To answer the questions as asked:

> Maybe sign_flags and enc_flags can be merged into a single flags parameter, used for both?
Yet I don't have sufficient experience with the CMS API whether this makes sense.
@slontis or @mattcaswell what do you think?

I think there could be scenarios where you want to control these separately. For example both the sign and encrypt APIs will accept the `CMS_TEXT` flag. But it seems to me to make little sense to use this on the internal sign operation, although you might conceivably want to use it on the outer encrypt operation.

> BTW, is the proposed name CMS_sign_encrypt() still acceptable (for consistency with existing CMS functions)?
Or do we feel obliged to use OSSL_CMS_sign_encrypt()?

Where we have a legacy established "namespace" then we can continue to use it. So new `CMS*` functions are fine in principle.

Having said all of that I have some reservations about the new proposed CMS API. I suppose this is being introduced as a convenience if you always want to both sign and encrypt data. But we already supply the sign and encrypt building blocks - so is this really necessary? I would prefer to avoid adding new API surface if we can avoid it.

A second concern is that this does sign first and then encrypt second. This reminds me of the "mac-then-encrypt" vs "encrypt-then-mac" issue in TLS. Originally TLS calculated a MAC on the application data first, and then encrypted the combined app data and MAC. Later it was decided that this was a bad idea and led to various attacks. This resulted in encrypt-then-mac where we did things the other way around (so encrypt first and then MAC the encrypted data). This meant, when receiving data, we always checked the integrity of the data as the first thing we do before attempting other operations. So, with this new API (if we even need it), would it be better to do things the other way around (encrypt first, sign second).

**Comment by DDvO** [comment]
Thanks a lot Matt for your very swift and helpful answers and comments!

Regarding the combination of flags, you brought up an interesting example.
Yet  `CMS_TEXT` could be handled in a smart way when combining the two flags into one,
namely by masking it for one of the two operations if it makes sense only for the other operation.
Are there other cases that may require special handling?

Yes, the combined sign&encrypt function is rather straightforward,
so it may be argued that it is not important to offer that function at the API level.
Instead, we may convert it to a function local to libcrypto - what do others think?

Very good point regarding whether one should really first sign data and then encrypt the result -
in fact, according to https://datatracker.ietf.org/doc/html/rfc9480#section-2.7 it should be the other way round!

**Comment by DDvO** [comment]
On a closer look again at RFC 9840, I have the impression that the text may just be misunderstood in the way I just did.

Also CMS [[RFC5652](https://datatracker.ietf.org/doc/html/rfc5652)] and [[RFC8933](https://datatracker.ietf.org/doc/html/rfc8933)] seems to suggest that when combining signed-data and enveloped-data, the signature is done first.

**Comment by mattcaswell** [comment]
> Yet CMS_TEXT could be handled in a smart way when combining the two flags into one,
namely by masking it for one of the two operations if it makes sense only for the other operation.
Are there other cases that may require special handling?

I haven't looked at all the flags. But what about `CMS_DETACHED`?

**Comment by mattcaswell** [comment]
> Yes, the combined sign&encrypt function is rather straightforward,
so it may be argued that it is not important to offer that function at the API level.
Instead, we may convert it to a function local to libcrypto - what do others think?

My tendency is to avoid new public API unless there's a really strong case for it. So my preference is to keep it local (we can always change our mind later...but we can't do it the other way around).

**Comment by slontis** [comment]
yes, normally you want to verify the signature before decrypting.

**Comment by t8m** [comment]
> yes, normally you want to verify the signature before decrypting.

This is unfortunately not how the CMS, S/MIME standards are spec-ed.

**Comment by DDvO** [comment]
Indeed. 
https://datatracker.ietf.org/doc/html/rfc8551#section-3.7 (S/MIME) says:

> It is possible to apply any of the signing, encrypting, and
   compressing operations in any order.  It is up to the implementer and
   the user to choose. 

And the CMS RFCs [5652](https://datatracker.ietf.org/doc/html/rfc5652) and [8933](https://datatracker.ietf.org/doc/html/rfc8933) apparently not give any caveat on the order of signing and encrypting.

**Comment by DDvO** [comment]
> yes, normally you want to verify the signature before decrypting.

Note that in CMP context the encrypted contents being received must be authenticated at outer message level,
so in fact there is a signature nested in an encryption nested in a signature.
From this I conclude on the CMP context it is fine.

**Comment by DDvO** [comment]
> My tendency is to avoid new public API unless there's a really strong case for it. So my preference is to keep it local (we can always change our mind later...but we can't do it the other way around).

Good point - so I'll add to the review a comment that the function should made internal.

**Comment by mattcaswell** [comment]
> I believe this option makes sense only for signatures, so it simply could be masked for the encryption part.

It looks to be supported for encryption too:

https://github.com/openssl/openssl/blob/7f62adaf2b088de38ad2e534d0bfae2ff7ae01f2/crypto/cms/cms_smime.c#L640-L676

**Comment by bukka** [comment]
>  BTW, is the proposed name CMS_sign_encrypt() still acceptable (for consistency with existing CMS functions)?
Or do we feel obliged to use OSSL_CMS_sign_encrypt()?

I think people should be moving away from sing and encrypt towards the auth enveloped data with key trans or key agree recipient info.

This is actually even mentioned in SMIME/4 spec: https://datatracker.ietf.org/doc/html/rfc8551#section-6
```
   Modification of the ciphertext in EnvelopedData can go undetected if
   authentication is not also used, which is the case when sending
   EnvelopedData without wrapping it in SignedData or enclosing
   SignedData within it.  This is one of the reasons for moving from
   EnvelopedData to AuthEnvelopedData, as the authenticated encryption
   algorithms provide the authentication without needing the SignedData
   layer.
```

So I'm not sure if it makes much sense to introduce a new function for something that people should really stop using for the new applications.
